### PR TITLE
feat: add support for function reference proposal for validation.

### PR DESCRIPTION
**Pull request by little-willy**
After #2151

### CODE REVIEW

**Comment by q82419** [comment]
Please fix the DCO.

**Comment by juntao** [comment]
Flows summarize

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: lib/validator/formchecker.cpp [hunk 1]
```diff
@@ -446,6 +516,17 @@ Expect<void> FormChecker::checkInstr(const AST::Instruction &Instr) {
     }
     return StackTrans(Types[N].first, Types[N].second);
   }
+  case OpCode::Call_ref: {
+    auto TypeIdx = Instr.getTargetIndex();
+    if (TypeIdx >= (uint32_t)Types.size()) {
```
Comment (q82419) [inline lib/validator/formchecker.cpp:521]: use `static_cast<uint32_t>`.

File: lib/validator/formchecker.cpp [hunk 2]
```diff
@@ -446,6 +516,17 @@ Expect<void> FormChecker::checkInstr(const AST::Instruction &Instr) {
...
+      return logOutOfRange(ErrCode::Value::InvalidFuncTypeIdx,
+                           ErrInfo::IndexCategory::FunctionType, TypeIdx,
+                           Types.size());
```
Comment (q82419) [inline lib/validator/formchecker.cpp:524]: use `static_cast<uint32_t>`.

**Review by q82419 — CHANGES_REQUESTED** [review]
Please help to rebase on the branch `proposal/func-ref`. Thanks.

File: lib/validator/formchecker.cpp [hunk 3]
```diff
@@ -492,6 +583,28 @@ Expect<void> FormChecker::checkInstr(const AST::Instruction &Instr) {
...
+      // Call function index out of range
+      spdlog::error(ErrCode::Value::InvalidFuncIdx);
+      spdlog::error(ErrInfo::InfoForbidIndex(
+          ErrInfo::IndexCategory::FunctionType, TypeIdx,
+          static_cast<uint32_t>(Types.size())));
+      return Unexpect(ErrCode::Value::InvalidFuncIdx);
+    }
+    if (!match_type(Types[TypeIdx].second, Returns)) {
+      spdlog::error(ErrCode::Value::TypeCheckFailed);
+      // TODO: Print the error info of types.
```
Comment (q82419) [inline lib/validator/formchecker.cpp:598]: Use `ErrInfo::InfoMismatch` instead of this TODO comment.

**Review by q82419 — CHANGES_REQUESTED** [review]

File: lib/validator/formchecker.cpp (hunk 3 above)
Comment (little-willy) [reply lib/validator/formchecker.cpp]: fixed.

**Pull request merged**